X509_check_host detection always fails

Configuration process fails to detect X509_check_host function even on systems that have it defined.

## Steps to reproduce

 1. Ensure we have proper verison of OpenSSL using `pkg-config --modversion libssl` (should return `1.1.0` or `1.0.2`).
 2. Clone the repository.
 3. Run `./autogen.sh`.
 4. Run `./configure.sh`.
 5. One of the last lines in the output is `checking for X509_check_host... no`.

This was tested on Fedora 27 (OpenSSL 1.1.0h) and unstable Gentoo (OpenSSL 1.0.2o).


## What is causing the issue

Main reason for this issue is the fact that PKG_CHECK_MODULES does not add proper linker flags to the set of flags that configure is using to check for function presence. This can be seen in the config.log file:

    configure:5102: checking for X509_check_host
    configure:5102: gcc -o conftest -g -O2   conftest.c -lutil -lpthread  >&5
    /tmp/cckiBJJz.o: In function `main':
    /home/tadej/devel/openfortivpn/conftest.c:184: undefined reference to `X509_check_host'
    collect2: error: ld returned 1 exit status

In the second line, `-lssl` and `-lcrypto` linker flags are missing.


## How to fix this

Initial stab at fixing this issue in in pull request #303

It looks like _autotools_ are a little bit broken/inconsistent here:
* _PKG_CHECK_MODULES_() does set/report a few variables:
```
pkg_cv_OPENSSL_CFLAGS=
pkg_cv_OPENSSL_LIBS='-lcrypto -lssl'
```
```
OPENSSL_CFLAGS=''
OPENSSL_LIBS='-lcrypto -lssl'
```
* _AC_CHECK_FUNCS_() does not use them:
```
configure:5114: gcc -o conftest -g -O2   conftest.c -lutil -lpthread  >&5
/tmp/cc4iZwHK.o: In function `main':
/volatile/src/openfortivpn/conftest.c:184: undefined reference to `X509_check_host'
collect2: error: ld returned 1 exit status
```

--- a/configure.ac
+++ b/configure.ac
@@ -18,7 +18,7 @@ AC_CANONICAL_HOST
 AM_SILENT_RULES([yes])
 
 # Checks for libraries.
-PKG_CHECK_MODULES(OPENSSL, [libcrypto libssl])
+PKG_CHECK_MODULES(OPENSSL, [libcrypto >= 0.9.8 libssl >= 0.9.8], [], [AC_MSG_ERROR([Cannot find OpenSSL 0.9.8 or higher.])])
 AC_CHECK_LIB([pthread], [pthread_create], [], [AC_MSG_ERROR([Cannot find libpthread.])])
 AC_CHECK_LIB([util], [forkpty], [], [AC_MSG_ERROR([Cannot find libutil.])])
 
@@ -39,6 +39,14 @@ AC_TYPE_UINT8_T
 AC_FUNC_MALLOC
 AC_FUNC_REALLOC
 AC_CHECK_FUNCS([atoi close connect execv exit _exit fclose fcntl fflush fopen forkpty fprintf fputs free freeaddrinfo freeifaddrs freopen fwrite getaddrinfo getchar getenv getopt_long htons index inet_addr inet_ntoa isatty malloc memcpy memmem memmove memset ntohs open openlog pclose popen printf pthread_cancel pthread_cond_init pthread_cond_signal pthread_cond_wait pthread_join pthread_mutexattr_init pthread_mutex_destroy pthread_mutex_init pthread_mutex_lock pthread_mutex_unlock pthread_sigmask puts read realloc rewind select setenv sigaddset sigemptyset signal snprintf socket sprintf strcasestr strcat strchr strcmp strcpy strdup strerror strlen strncasecmp strncat strncpy strsignal strstr strtok strtok_r strtol syslog system tcsetattr usleep vprintf vsnprintf vsyslog write], [], AC_MSG_ERROR([Required function not present]))
-AC_CHECK_FUNCS([pthread_mutexattr_setrobust X509_check_host])
+AC_CHECK_FUNCS([pthread_mutexattr_setrobust])
+# Use PKG_CHECK_MODULES compiler/linker flags
+save_openssl_CPPFLAGS="${CPPFLAGS}"
+save_openssl_LIBS="${LIBS}"
+CPPFLAGS="${OPENSSL_CFLAGS} ${CPPFLAGS}"
+LIBS="${OPENSSL_LIBS} ${LIBS}"
+AC_CHECK_FUNCS([X509_check_host])
+CPPFLAGS="${save_openssl_CPPFLAGS}"
+LIBS="${save_openssl_LIBS}"
 
 AC_OUTPUT(Makefile)

